src/ucs/datastruct/ptr_array.c

```diff
@@ -172,37 +180,126 @@ unsigned ucs_ptr_array_insert(ucs_ptr_array_t *ptr_array, void *value,
 
-    /* Populate */
-    *placeholder_p = ucs_ptr_array_placeholder_get(*elem);
-    *elem = (uintptr_t)value;
-    return elem_index;
+    return element_index;
 }
 
-void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned elem_index,
-                          uint32_t placeholder)
+void ucs_ptr_array_remove(ucs_ptr_array_t *ptr_array, unsigned element_index)
 {
-    ucs_ptr_array_elem_t *elem = &ptr_array->start[elem_index];
+    ucs_ptr_array_elem_t *elem = &ptr_array->start[element_index];
+    ucs_ptr_array_elem_t *next_elem;
+    uint32_t size_free_ahead;
+
+    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, element_index));
+
+    if (ucs_ptr_array_is_free(ptr_array, element_index + 1)) {
+        next_elem = &ptr_array->start[element_index + 1];
+        size_free_ahead = ucs_ptr_array_size_free_get(*next_elem);
+        size_free_ahead++;
+    } else {
+        size_free_ahead = 1;
+    }
+
+    ucs_ptr_array_freelist_element_set(elem, size_free_ahead,
+                                       ptr_array->freelist);
 
-    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, elem_index));
-    *elem = UCS_PTR_ARRAY_FLAG_FREE;
-    ucs_ptr_array_placeholder_set(elem, placeholder);
-    ucs_ptr_array_freelist_set_next(elem, ptr_array->freelist);
-    ptr_array->freelist = elem_index;
+    /* Make sure the next element is free */
+    ucs_assert(__ucs_ptr_array_is_free(ptr_array->start[element_index + size_free_ahead - 1]));
+
+    ptr_array->freelist = element_index;
 }
 
-void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned elem_index, void *new_val)
+void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned element_index,
+                            void *new_val)
 {
     void *old_elem;
 
-    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, elem_index));
-    old_elem = (void *)ptr_array->start[elem_index];
-    ptr_array->start[elem_index] = (uintptr_t)new_val;
+    ucs_assert_always(!ucs_ptr_array_is_free(ptr_array, element_index));
+    old_elem = (void *)ptr_array->start[element_index];
+    ptr_array->start[element_index] = (uintptr_t)new_val;
     return old_elem;
 }
+
+
+/*
+ *  Locked interface functions implementation
+ */
+
+ucs_status_t
+ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array,
+                          const char *name)
+{
+    ucs_status_t status;
+
+    /* Initialize spinlock */
+    status = ucs_recursive_spinlock_init(&locked_ptr_array->lock, 0);
+    if (status != UCS_OK) {
+       return status;
+    }
+
+    /* Call unlocked function */
+    ucs_ptr_array_init(&locked_ptr_array->super, name);
+
+    return UCS_OK;
+}
+
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array)
+{
+    ucs_status_t status;
+
+    ucs_recursive_spin_lock(&locked_ptr_array->lock);
+    /* Call unlocked function */
+    ucs_ptr_array_cleanup(&locked_ptr_array->super);
+    ucs_recursive_spin_unlock(&locked_ptr_array->lock);
```

Comment:
doesn't make sense to stake lock during cleanup, cleanup is not thread safe anyway